sdks/java/io/elasticsearch/common/src/main/java/org/apache/beam/sdk/io/elasticsearch/common/ElasticsearchIOCore.java

```diff
@@ -75,104 +72,65 @@
 import org.elasticsearch.client.RestClientBuilder;
 
 /**
- * Transforms for reading and writing data from/to Elasticsearch.
- * This IO is only compatible with Elasticsearch v2.x
- *
- * <h3>Reading from Elasticsearch</h3>
- *
- * <p>{@link ElasticsearchIO#read ElasticsearchIO.read()} returns a bounded
- * {@link PCollection PCollection&lt;String&gt;} representing JSON documents.
- *
- * <p>To configure the {@link ElasticsearchIO#read}, you have to provide a connection configuration
- * containing the HTTP address of the instances, an index name and a type. The following example
- * illustrates options for configuring the source:
- *
- * <pre>{@code
- *
- * pipeline.apply(ElasticsearchIO.read().withConnectionConfiguration(
- *    ElasticsearchIO.ConnectionConfiguration.create("http://host:9200", "my-index", "my-type")
- * )
- *
- * }</pre>
- *
- * <p>The connection configuration also accepts optional configuration: {@code withUsername()} and
- * {@code withPassword()}.
- *
- * <p>You can also specify a query on the {@code read()} using {@code withQuery()}.
- *
- * <h3>Writing to Elasticsearch</h3>
- *
- * <p>To write documents to Elasticsearch, use
- * {@link ElasticsearchIO#write ElasticsearchIO.write()}, which writes JSON documents from a
- * {@link PCollection PCollection&lt;String&gt;} (which can be bounded or unbounded).
- *
- * <p>To configure {@link ElasticsearchIO#write ElasticsearchIO.write()}, similar to the read, you
- * have to provide a connection configuration. For instance:
- *
- * <pre>{@code
- *
- *  pipeline
- *    .apply(...)
- *    .apply(ElasticsearchIO.write().withConnectionConfiguration(
- *       ElasticsearchIO.ConnectionConfiguration.create("http://host:9200", "my-index", "my-type")
- *    )
- *
- * }</pre>
- *
- * <p>Optionally, you can provide {@code withBatchSize()} and {@code withBatchSizeBytes()}
- * to specify the size of the write batch in number of documents or in bytes.
+ * Core of ElasticsearchIO (common code for all the versions).
+ * This class should not be used by directly.
  */
-@Experimental(Experimental.Kind.SOURCE_SINK)
-public class ElasticsearchIO {
+public class ElasticsearchIOCore {
+
+  private static ElasticsearchRequester requester;
+
+  public static void setRequester(ElasticsearchRequester r){
```

Comment:
I don't like it either. As I wrote above, the composition design pattern is not ideal. Requester is static because it is used by all the static nested classes. Please note that, as I explained above, all the common classes are public because they are used in version specific packages but they shall not be used by the user, setRequester() will be called only by one ElasticsearchIO so with always the same argument. Can you suggest a way to improve this design?